app_tests/integration_tests/llm/shortfin/tracing_test.py

```diff
@@ -0,0 +1,153 @@
+
+logger = logging.getLogger(__name__)
+
+pytestmark = pytest.mark.parametrize(
+    "model_artifacts,server",
+    [
+        ["tinystories_llama2_25m", {"prefix_sharing": "none"}],
+    ],
+    indirect=True,
+)
+
+# Same golden values as in tinystories_llama2_25m_test.py
+GOLDEN_PROMPT = "Once upon a time"
+GOLDEN_RESPONSE = ", there was a little girl named Lily."  # this assumes purely deterministic greedy search
+
+
+class TestTracingFunctionality:
+    """Test suite for tracing functionality with the LLM server."""
+
+    def test_tracing_generation(self, server: Tuple[Any, int], tmp_path: Path) -> None:
+        """Tests that tracing works while running text generation.
+
+        Args:
+            server: Tuple of (process, port) from server fixture
+            tmp_path: Temporary directory for storing the trace file
+        """
+        process, port = server
+        assert process.poll() is None, "Server process terminated unexpectedly"
+
+        # Determine output directory for artifacts (use environment variable if set, otherwise tmp_path)
+        output_dir = os.environ.get("TEST_OUTPUT_DIR")
+        if output_dir:
+            output_dir = Path(output_dir)
+            output_dir.mkdir(parents=True, exist_ok=True)
+        else:
+            output_dir = tmp_path
+
+        # Start tracy capture
+        capture_file = output_dir / "capture.tracy"
+        logger.info(f"Tracy capture will be saved to: {capture_file}")
+
+        tracy_process = subprocess.Popen(
+            ["iree-tracy-capture", "-o", str(capture_file)],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+
+        try:
+            # Wait a moment for tracy capture to initialize
+            time.sleep(3)
+
+            # Run generation with tracing enabled
+            prompt = GOLDEN_PROMPT
+            expected_prefix = GOLDEN_RESPONSE
+
+            # Set environment variable to enable tracing for the request if not already set
+            tracy_env_was_set = "SHORTFIN_PY_RUNTIME" in os.environ
+            if not tracy_env_was_set:
+                os.environ["SHORTFIN_PY_RUNTIME"] = "tracy"
+
+            # Make generation request
+            response = self._generate(prompt, port)
+
+            # Verify the response is correct to ensure basic functionality works
+            if not expected_prefix in response:
+                raise AccuracyValidationException(
+                    expected=f"{expected_prefix}...",
+                    actual=response,
+                    message=f"Generation did not match expected pattern.\nExpected to start with: {expected_prefix}\nActual response: {response}",
+                )
+
+            # Clean up environment variable if we set it
+            if not tracy_env_was_set:
+                os.environ.pop("SHORTFIN_PY_RUNTIME", None)
+
+            # Wait a moment for trace data to be collected
+            time.sleep(5)
```

Comment:
Same here. You might want to set `TRACY_NO_EXIT`. See also https://iree.dev/developers/performance/profiling-with-tracy/#quickstart. I would hope you could remove all the timeouts and waits from this test and just rely on pytest-timeout.

Another thing to consider will be having multiple test processes or user processes using the same ports. May want to be explicit with each test and the capture tool so ensure no overlap.